web/libs/editor/src/components/VideoCanvas/VideoCanvas.tsx

```diff
@@ -264,11 +264,13 @@ export const VideoCanvas = memo(
       updateFrame();
 
       if (playing) {
-        raf.current = requestAnimationFrame(handleAnimationFrame);
+        raf.current = requestAnimationFrame(handleAnimationFrameRef.current);
       } else {
         cancelAnimationFrame(raf.current!);
       }
     };
+    const handleAnimationFrameRef = useRef(handleAnimationFrame);
+    handleAnimationFrameRef.current = handleAnimationFrame;
```

Comment:
this looks so confusing... are there other ways to have recursion in functional components?

Reply:
Recursion works fine. It just can't use the right values inside the function without moving them to ref if they are changing during execution.
It's the same trick as when we move handlers from props to ref to decrease the necessity to invalidate caches of useMemo/useCallback.